### ISSUE TITLE
Potential bug with diag

### ISSUE THREAD

**Issue by howardnewyork** [greta-dev/greta#216]
diag function does not seem to be working with greta arrays, e.g.:

N=100
xx=as_data(1)
dim(diag(xx,N,N))

produces
`[1] 1 1`

but should be 
`[1] 100 100`

**Comment by goldingn** [comment]
Yeah, `diag()` can't be used to create greta arrays yet, only to extract or replace elements. That's documented [here](https://greta-dev.github.io/greta/functions.html), though it should error with a helpful message. Its also an annoying thing not to have and shouldn't be difficult to fix.

I think this is a duplicate of #108, but this has more info. So I'll close that one. Thanks!

**Comment by Voltemand** [comment]
I have a branch on my fork which fixes this. I insert:

```{r}
# if the greta_array is like a vector then construct a greta_array matrix
if (length(dim[2]) == 1) {
  vec <- as.vector(get_node(x)$.value)
  out <- base::diag(vec, nrow, ncol)
  return(as.greta_array(out))
}
```
into the `diag.greta_array` method.  I'm not sure if this is the best way to do this though.

**Comment by goldingn** [comment]
Nice, thanks!

I'm hoping to cover all of the different ways of constructing a diagonal matrix with diag, which, from memory, also includes creating non-square diagonal matrices.

When I get to that I'll try to include your fork though. Could you please point me to it?

**Comment by Voltemand** [comment]
My fork is Voltemand/greta/issue216 (I have now stopped naming my branches after issues as I forget what they are all about!) I basically insert the following code into `diag.greta_array`:

```R
# try to construct a greta array if either nrow or ncol are specified
  if (!missing(nrow) | !missing(ncol)) {

      if (dim[2] != 1) {
          stop("'nrow' or 'ncol' cannot be specified when 'x' is not a vector",
               " like greta_array")

      } else {
          vec <- as.vector(get_node(x)$value())
          out <- diag(vec, nrow, ncol)
          return(as.greta_array(out))
      }
  }
```
I think (hope?) that the use of `diag` again (possibly it should be `base::diag`?) covers the cases you mentioned. 

All the best!

**Comment by goldingn** [comment]
Great, thanks! I'll write up some tests to cover all the documented behaviours and give this a go.
